### DISCUSSION TITLE
`pre-commit` runs fine from command line `git commit`, not from magit.

### DISCUSSION THREAD

**Discussion by joostkremers** [magit/magit#4654] [Support] [answered]
`magit-version` is `"20220406.1950"`, `emacs-version` is `"28.1.50"`.

I have a pre-commit hook in one of my repos that checks whether the markdown source for the documentation of the package in question has changed, and if so, regenerates the info and HTML versions of said documentation. This hook works fine when I run `git commit` from the command line, and it used to work fine when I committed changes to the documentation using magit.

For some reason, however, it no longer works when I try to commit from magit. Instead, an empty commit is created and the magit-process buffer reports the following error: `pathspec ':/' did not match any file(s) known to git`. Committing from the command line still works without problems, however.

This is the relevant part of the `pre-commit` hook (the complete script can be seen [here](https://github.com/joostkremers/ebib/blob/master/pre-commit.sh)):

``` shell
if [ "$(git status --porcelain manual/ebib.text)" == 'M  manual/ebib.text' ] ; then
    echo "$SCRIPT: regenerating manual files"
    git stash -q --keep-index
    create_texi "$SOURCE" && run_makeinfo "$TEXINFO" && git add "$INFO"
    create_html "$SOURCE" && git add "$HTML"
    git stash pop -q
fi
```

This is the output in the magit-process buffer:

```
pre-commit: regenerating manual files
error: pathspec ':/' did not match any file(s) known to git
pre-commit: running pandoc to create texinfo
pre-commit: runnig makeinfo
pre-commit: running pandoc to create HTML
pre-commit: documentation is current
hint: Waiting for your editor to close the file...
Waiting for Emacs...
[master e9c29bc] test
```

The line in the `pre-commit` hook causing the error is apparently `git stash -q --keep-index`, since the `create_texi` function is the one that prints `"pre-commit: running pandoc to create texinfo"`.

Any idea what might be going on here? Like I said, doing the commit from the command line is unproblematic, and so is a `git stash -q --keep-index` on the command line.

BTW, in the instructions for new issues, it seems that `magit-section-N` is missing from this command:

    $ cd ~/.emacs.d/elpa/magit-N
    $ emacs -Q --debug-init --eval '(setq debug-on-error t)' -L ../dash-N -L ../git-commit-N -L ../transient-N -L ../with-editor-N -L ../ghub-N -L . -l magit

(The problem described above occurs with `emacs -Q`. :slightly_smiling_face: )

**Comment by tarsius** [comment, marked as answer]
While it is possible that Magit should do something differently, I think it is more likely that something with your scripts is wrong. So I am turning this into a support discussion.

> The line in the `pre-commit` hook causing the error is apparently `git stash -q --keep-index`, since the `create_texi` function is the one that prints `"pre-commit: running pandoc to create texinfo"`.

As far as I can tell it could also be something that comes later. ~~Then again, I don't know what happens inside `create_texi`.~~

I would start by printing the value of `$SOURCE` and all the other variables too.

**Reply by joostkremers** [reply]
So I added some code to the pre-commit hook to print the variable values (see https://github.com/joostkremers/ebib/commit/2c20a8c5852c36c9631d555b52b7ef2a2333b1c0). Then I made some changes to the documentation source, staged them and tried to commit. The magit-process buffer now says:

```
SOURCE: manual/ebib.text
TEXINFO: manual/ebib.texi
INFO: ebib.info
HTML: docs/ebib-manual.html
pre-commit: regenerating manual files
error: pathspec ':/' did not match any file(s) known to git
pre-commit: running pandoc to create texinfo
pre-commit: runnig makeinfo
pre-commit: running pandoc to create HTML
pre-commit: documentation is current
hint: Waiting for your editor to close the file...
Waiting for Emacs...
[pre-commit e80c9ae] test
```

And an empty commit was created (see https://github.com/joostkremers/ebib/commit/e80c9aef51d8e11fa7b5a7f81c4079fdcff25069).

After the commit, the changes I had made and which I had staged before, were now unstaged again.

At this point, doing on the command line `git add manual/ebib.text` followed by `git commit` ran the pre-commit hook successfully and created the commit:

```
joost@Swift:~/src/Elisp/ebib$ git add manual/ebib.text 
joost@Swift:~/src/Elisp/ebib$ git commit 
SOURCE: manual/ebib.text
TEXINFO: manual/ebib.texi
INFO: ebib.info
HTML: docs/ebib-manual.html
pre-commit: regenerating manual files
pre-commit: running pandoc to create texinfo
pre-commit: runnig makeinfo
pre-commit: running pandoc to create HTML
pre-commit: documentation is current
hint: Waiting for your editor to close the file... Waiting for Emacs...
[pre-commit 24efd86] Now committing from the command line.
 3 files changed, 148 insertions(+), 147 deletions(-)
```

The commit can be seen here: https://github.com/joostkremers/ebib/commit/24efd86dc999dbd629c9e7d05482fac8e8216f4d.

**Reply by kyleam** [reply]
This is a Git bug.  It bisects to b932f6a5e8 (stash: fix handling removed files with --keep-index, 2019-07-16), part of the v2.23.0 release.

I'll see if I can come up with a patch, but either way, I'll report it to the Git list within the next couple of days.

<details>
<summary>bisect script</summary>

```sh
#!/bin/sh

set -eu

make -j3 || exit 125

PATH=$PWD//bin-wrappers:$PATH
export PATH

tdir=$(mktemp -d "${TMPDIR:-/tmp}"/git-XXXXXXX)
cd "$tdir"

git init
echo foo >foo
echo bar >bar
git add .
git commit -mc0

echo more >>foo
echo more >>bar

git add foo

git --literal-pathspecs stash --keep-index
```

</details>

**Reply by kyleam** [reply]
In the meantime, I'd guess `s/git stash -q --keep-index/git --no-literal-pathspecs stash -q --keep-index/` in your script would resolve the issue.

Edit: Or you could try going with the manual's suggestion and adding `unset GIT_LITERAL_PATHSPECS` in your script.

https://magit.vc/manual/magit/My-Git-hooks-work-on-the-command_002dline-but-not-inside-Magit.html#My-Git-hooks-work-on-the-command_002dline-but-not-inside-Magit

**Reply by kyleam** [reply]
> I'll see if I can come up with a patch [...]

Sent: https://lore.kernel.org/git/20220408031228.782547-1-kyle@kyleam.com/

**Reply by joostkremers** [reply]
Thanks! It never occurred to me to check the Magit FAQ... (Though perhaps if this is indeed a Git bug, then something good came out of that oversight.)

Anyway, I can confirm that unsetting `GIT_LITERAL_PATHSPECS` in my pre-commit hook works.
